packages/dev/core/src/assetContainer.ts

```diff
@@ -116,6 +116,110 @@ export class AssetContainer extends AbstractScene {
+
+        // First pass: Initialize the empty dependency graph
+        for (const node of nodes) {
+            const nodeId = node.uniqueId;
+            dependencyGraph.dependsOn.set(nodeId, new Set<number>());
+            dependencyGraph.dependedBy.set(nodeId, new Set<number>());
+        }
+
+        // Second pass: Populate the dependency graph. We assume that we
+        // don't need to check for cycles here, as the scene graph cannot
+        // contain cycles. Our graph also already contains all transitive
+        // dependencies because getDescendants returns the transitive
+        // dependencies by default.
+        for (const node of nodes) {
+            const nodeId = node.uniqueId;
+            const dependsOn = dependencyGraph.dependsOn.get(nodeId)!;
+            if (node instanceof InstancedMesh) {
+                const masterMesh = node.sourceMesh;
+                if (nodesUidMap.has(masterMesh.uniqueId)) {
+                    dependsOn.add(masterMesh.uniqueId);
+                    dependencyGraph.dependedBy.get(masterMesh.uniqueId)!.add(nodeId);
+                }
+            }
+            const dependedBy = dependencyGraph.dependedBy.get(nodeId)!;
+
+            for (const child of node.getDescendants()) {
+                const childId = child.uniqueId;
+                if (nodesUidMap.has(childId)) {
+                    dependedBy.add(childId);
+
+                    const childDependsOn = dependencyGraph.dependsOn.get(childId)!;
+                    childDependsOn.add(nodeId);
+                }
+            }
+        }
+
+        // Third pass: Topological sort
+        const sortedNodes: Node[] = [];
+        const nodesMarkedForVisit = new Set<number>();
+
+        // First: Find all nodes that have no dependencies
+        const leaves: Node[] = [];
+        for (const node of nodes) {
+            const nodeId = node.uniqueId;
+            if (dependencyGraph.dependsOn.get(nodeId)!.size === 0) {
+                leaves.push(node);
+            }
+        }
+
+        const visitList = leaves;
+        while (visitList.length > 0) {
+            const nodeToVisit = visitList.splice(0, 1)[0];
+
+            sortedNodes.push(nodeToVisit);
+
+            // Remove the node from the dependency graph
+            // When a node is visited, we know that dependsOn is empty.
+            // So we only need to remove the node from dependedBy.
+            const dependedByVisitedNode = dependencyGraph.dependedBy.get(nodeToVisit.uniqueId)!;
+            for (const dependedByVisitedNodeId of dependedByVisitedNode) {
+                const dependsOnDependedByVisitedNode = dependencyGraph.dependsOn.get(dependedByVisitedNodeId)!;
+                dependsOnDependedByVisitedNode.delete(nodeToVisit.uniqueId);
+
+                if (dependsOnDependedByVisitedNode.size === 0 && !nodesMarkedForVisit.has(dependedByVisitedNodeId)) {
+                    visitList.push(nodesUidMap.get(dependedByVisitedNodeId)!);

```

Comment:
Instead of using a new `nodesMarkedForVisit` set, maybe you can reuse `nodesUidMap` instead and do something like:
```typescript
 if (dependsOnDependedByVisitedNode.size === 0 && nodesUidMap.get(dependedByVisitedNodeId)) {
    visitList.push(nodesUidMap.get(dependedByVisitedNodeId)!);
    nodesUidMap.delete(dependedByVisitedNodeId);
}
```
As an added bonus, at the end you can issue a warning/error if `nodesUidMap` is not empty, which would be an easy way to detect problems in the graph (you can even output the names of problematic nodes if you want).